File: sdks/go/src/core/config_file.go
package core

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// FileConfig is the YAML-serializable representation of SDK configuration.
// It maps to the loza.yaml file format.
type FileConfig struct {
	CollectorURL      string `yaml:"collector_url"`
	ServiceName       string `yaml:"service_name"`
	ServiceVersion    string `yaml:"service_version"`
	Environment       string `yaml:"environment"`
	TenantID          string `yaml:"tenant_id"`
	BatchSize         int    `yaml:"batch_size"`
	FlushInterval     string `yaml:"flush_interval"`
	MaxBufferSize     int    `yaml:"max_buffer_size"`
	MaxRetries        int    `yaml:"max_retries"`
	MaxBackoff        string `yaml:"max_backoff"`
	Timeout           string `yaml:"timeout"`
	ConnectionTimeout string `yaml:"connection_timeout"`
	EnableCompression *bool  `yaml:"enable_compression"`
}

// ErrConfigFileNotFound is returned when no config file is found.
var ErrConfigFileNotFound = errors.New("loza: config file not found")

// LoadFromFile loads configuration from a loza.yaml file.
// If path is empty, it searches for loza.yaml in the current directory,
// then in the user's home directory (~/.loza/loza.yaml).
//
// Returns ErrConfigFileNotFound if no config file is found.
// Returns a parse error if the file exists but cannot be parsed.
//
// This implements Requirement 32.3.
func LoadFromFile(path string) (FileConfig, error) {
	if path == "" {
		path = findConfigFile()
	}
	if path == "" {
		return FileConfig{}, ErrConfigFileNotFound
	}

	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return FileConfig{}, ErrConfigFileNotFound
		}
		return FileConfig{}, fmt.Errorf("loza: read config file %q: %w", path, err)
	}

	var fc FileConfig
	if err := parseYAML(data, &fc); err != nil {
		return FileConfig{}, fmt.Errorf("loza: parse config file %q: %w", path, err)
	}
	return fc, nil
}

// LoadDefaultsFile loads repo-level SDK defaults from loza-go.defaults.yaml.
func LoadDefaultsFile() (FileConfig, error) {
	path := findDefaultsConfigFile()
	data, err := os.ReadFile(path)
	if err != nil {
		return FileConfig{}, fmt.Errorf("loza: read defaults file %q: %w", path, err)
	}

	var fc FileConfig
	if err := parseYAML(data, &fc); err != nil {
		return FileConfig{}, fmt.Errorf("loza: parse defaults file %q: %w", path, err)
	}
	return fc, nil
}

// findConfigFile searches for loza.yaml in standard locations.
func findConfigFile() string {
	// 1. Current directory
	if _, err := os.Stat("loza.yaml"); err == nil {
		return "loza.yaml"
	}
	// 2. Home directory
	if home, err := os.UserHomeDir(); err == nil {
		p := filepath.Join(home, ".loza", "loza.yaml")
		if _, err := os.Stat(p); err == nil {
			return p
		}
	}
	return ""
}

func findDefaultsConfigFile() string {
	if override := strings.TrimSpace(os.Getenv("LOZA_GO_DEFAULTS")); override != "" {
		return override
	}

	candidates := []string{
		"loza-go.defaults.yaml",
		filepath.Join("..", "loza-go.defaults.yaml"),
		filepath.Join("..", "..", "loza-go.defaults.yaml"),
	}
	for _, candidate := range candidates {
		if _, err := os.Stat(candidate); err == nil {
			return candidate
		}
	}

	if exe, err := os.Executable(); err == nil {
		dir := filepath.Dir(exe)
		for i := 0; i < 5; i++ {
			candidate := filepath.Join(dir, "loza-go.defaults.yaml")
			if _, statErr := os.Stat(candidate); statErr == nil {
				return candidate
			}
			dir = filepath.Dir(dir)
		}
	}
	return "loza-go.defaults.yaml"
}

// mergeFileConfig applies file-based config values to base, only overriding
// zero-value fields (file config has lower precedence than env vars).
func mergeFileConfig(base Config, fc FileConfig) Config {
	if fc.CollectorURL != "" && base.CollectorURL == "" {
		base.CollectorURL = fc.CollectorURL
	}
	if fc.ServiceName != "" && base.Service == "" {
		base.Service = fc.ServiceName
	}
	if fc.ServiceVersion != "" && base.Version == "" {
		base.Version = fc.ServiceVersion
	}
	if fc.Environment != "" && base.Environment == "" {
		base.Environment = fc.Environment
	}
	if fc.TenantID != "" && base.TenantID == "" {
		base.TenantID = fc.TenantID
	}
	if fc.BatchSize > 0 && base.BatchSize == 0 {
		base.BatchSize = fc.BatchSize
	}
	if fc.FlushInterval != "" && base.FlushInterval == 0 {
		if d, err := time.ParseDuration(fc.FlushInterval); err == nil {
			base.FlushInterval = d
		}
	}
	if fc.MaxBufferSize > 0 && base.MaxBufferSize == 0 {
		base.MaxBufferSize = fc.MaxBufferSize
	}
	if fc.MaxRetries > 0 && base.MaxRetries == 0 {
		base.MaxRetries = fc.MaxRetries
	}
	if fc.MaxBackoff != "" && base.MaxBackoff == 0 {
		if d, err := time.ParseDuration(fc.MaxBackoff); err == nil {
			base.MaxBackoff = d
		}
	}
	if fc.Timeout != "" && base.Timeout == 0 {
		if d, err := time.ParseDuration(fc.Timeout); err == nil {
			base.Timeout = d
		}
	}
	if fc.ConnectionTimeout != "" && base.ConnectionTimeout == 0 {
		if d, err := time.ParseDuration(fc.ConnectionTimeout); err == nil {
			base.ConnectionTimeout = d
		}
	}
	if fc.EnableCompression != nil {
		base.EnableCompression = *fc.EnableCompression
	}
	return base
}

func overlayFileConfig(base, override FileConfig) FileConfig {
	if override.CollectorURL != "" {
		base.CollectorURL = override.CollectorURL
	}
	if override.ServiceName != "" {
		base.ServiceName = override.ServiceName
	}
	if override.ServiceVersion != "" {
		base.ServiceVersion = override.ServiceVersion
	}
	if override.Environment != "" {
		base.Environment = override.Environment
	}
	if override.TenantID != "" {
		base.TenantID = override.TenantID
	}
	if override.BatchSize > 0 {
		base.BatchSize = override.BatchSize
	}
	if override.FlushInterval != "" {
		base.FlushInterval = override.FlushInterval
	}
	if override.MaxBufferSize > 0 {
		base.MaxBufferSize = override.MaxBufferSize
	}
	if override.MaxRetries > 0 {
		base.MaxRetries = override.MaxRetries
	}
	if override.MaxBackoff != "" {
		base.MaxBackoff = override.MaxBackoff
	}
	if override.Timeout != "" {
		base.Timeout = override.Timeout
	}
	if override.ConnectionTimeout != "" {
		base.ConnectionTimeout = override.ConnectionTimeout
	}
	if override.EnableCompression != nil {
		base.EnableCompression = override.EnableCompression
	}
	return base
}

// parseYAML parses YAML data into a FileConfig without requiring an external
// dependency. It uses a simple line-by-line parser for the flat key: value
// format used by loza.yaml.
//
// This avoids adding a yaml dependency to the core module. For production use,
// callers can use LoadFromFileWithYAML which accepts pre-parsed data.
func parseYAML(data []byte, fc *FileConfig) error {
	return parseSimpleYAML(string(data), fc)
}

// parseSimpleYAML parses a simple flat YAML file (key: value pairs only).
// This handles the loza.yaml format without requiring an external YAML library.
func parseSimpleYAML(content string, fc *FileConfig) error {
	lines := splitLines(content)
	for _, line := range lines {
		// Skip comments and empty lines
		trimmed := trimSpace(line)
		if trimmed == "" || trimmed[0] == '#' {
			continue
		}

		key, value, ok := splitKeyValue(trimmed)
		if !ok {
			continue
		}

		switch key {
		case "collector_url":
			fc.CollectorURL = value
		case "service_name":
			fc.ServiceName = value
		case "service_version":
			fc.ServiceVersion = value
		case "environment":
			fc.Environment = value
		case "tenant_id":
			fc.TenantID = value
		case "batch_size":
			if n, err := parseInt(value); err == nil {
				fc.BatchSize = n
			}
		case "flush_interval":
			fc.FlushInterval = value
		case "max_buffer_size":
			if n, err := parseInt(value); err == nil {
				fc.MaxBufferSize = n
			}
		case "max_retries":
			if n, err := parseInt(value); err == nil {
				fc.MaxRetries = n
			}
		case "max_backoff":
			fc.MaxBackoff = value
		case "timeout":
			fc.Timeout = value
		case "connection_timeout":
			fc.ConnectionTimeout = value
		case "enable_compression":
			b := value == "true" || value == "yes" || value == "1"
			fc.EnableCompression = &b
		}
	}
	return nil
}

func splitLines(s string) []string {
	var lines []string
	start := 0
	for i := 0; i < len(s); i++ {
		if s[i] == '\n' {
			lines = append(lines, s[start:i])
			start = i + 1
		}
	}
	if start < len(s) {
		lines = append(lines, s[start:])
	}
	return lines
}

func trimSpace(s string) string {
	start := 0
	for start < len(s) && (s[start] == ' ' || s[start] == '\t' || s[start] == '\r') {
		start++
	}
	end := len(s)
	for end > start && (s[end-1] == ' ' || s[end-1] == '\t' || s[end-1] == '\r') {
		end--
	}
	return s[start:end]
}

func splitKeyValue(s string) (key, value string, ok bool) {
	for i := 0; i < len(s); i++ {
		if s[i] == ':' {
			key = trimSpace(s[:i])
			value = trimSpace(s[i+1:])
			// Strip inline comments
			for j := 0; j < len(value); j++ {
				if value[j] == '#' {
					value = trimSpace(value[:j])
					break
				}
			}
			// Strip surrounding quotes
			if len(value) >= 2 && ((value[0] == '"' && value[len(value)-1] == '"') ||
				(value[0] == '\'' && value[len(value)-1] == '\'')) {
				value = value[1 : len(value)-1]
			}
			return key, value, key != ""
		}
	}
	return "", "", false
}

func parseInt(s string) (int, error) {
	n := 0
	if len(s) == 0 {
		return 0, fmt.Errorf("empty string")
	}
	for _, c := range s {
		if c < '0' || c > '9' {
			return 0, fmt.Errorf("invalid integer: %q", s)
		}
		n = n*10 + int(c-'0')
	}
	return n, nil
}

File: sdks/js/src/config/config-file.ts
/**
 * YAML-based config file loading for the JS/TS SDK.
 *
 * Implements the 4-layer config precedence:
 *   1. loza-js.defaults.yaml  (committed SDK defaults)
 *   2. .loza-js.yaml / loza.yaml  (user overrides)
 *   3. Environment variables (including LOZA_DSN)
 *   4. Code-level config (builder)
 *
 * Uses a simple built-in YAML parser (flat key: value + one level of nesting)
 * to avoid external dependencies.
 */

import { existsSync, readFileSync } from 'node:fs';
import { resolve, join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import type { Config } from './config.ts';
import { parse as parseDSN } from './dsn.ts';

// ── Simple YAML parser ──────────────────────────────────────────────────────

/** Parsed YAML as a nested plain object. */
export type YamlRecord = Record<string, unknown>;

/**
 * Parse simple YAML content into a plain object.
 * Handles flat key: value pairs and one level of nesting.
 * Comments (# ...) and blank lines are skipped.
 * Values are parsed as booleans, integers, or strings.
 */
export function parseSimpleYAML(content: string): YamlRecord {
  const root: YamlRecord = {};
  const stack: Array<{ indent: number; obj: YamlRecord }> = [{ indent: -1, obj: root }];

  for (const rawLine of content.split('\n')) {
    const trimmed = rawLine.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;

    const indent = rawLine.length - rawLine.trimStart().length;

    // Pop stack until we find the parent for this indent level
    while (stack.length > 1 && indent <= stack[stack.length - 1].indent) {
      stack.pop();
    }

    const colonIdx = trimmed.indexOf(':');
    if (colonIdx === -1) continue;

    const key = trimmed.slice(0, colonIdx).trim();
    const rawValue = trimmed.slice(colonIdx + 1).trim();

    // Strip inline comments
    let value = rawValue;
    const commentIdx = value.indexOf('#');
    if (commentIdx > 0) {
      value = value.slice(0, commentIdx).trim();
    }

    const current = stack[stack.length - 1].obj;

    if (value === '') {
      // Empty value means nested object
      const child: YamlRecord = {};
      current[key] = child;
      stack.push({ indent, obj: child });
    } else {
      current[key] = parseScalar(value);
    }
  }

  return root;
}

/** Parse a YAML scalar value (string, boolean, integer). */
function parseScalar(value: string): unknown {
  // Strip surrounding quotes
  if (
    value.length >= 2 &&
    value[0] === value[value.length - 1] &&
    (value[0] === '"' || value[0] === "'")
  ) {
    value = value.slice(1, -1);
  }

  const lower = value.toLowerCase();
  if (lower === 'true') return true;
  if (lower === 'false') return false;

  // Integer
  if (/^-?\d+$/.test(value)) {
    const n = parseInt(value, 10);
    if (!isNaN(n)) return n;
  }

  return value;
}

// ── Path resolution ─────────────────────────────────────────────────────────

/** Resolve the package root directory (two levels up from this file). */
function getPackageRoot(): string {
  try {
    const here = dirname(fileURLToPath(import.meta.url));
    return resolve(here, '..', '..');
  } catch {
    return process.cwd();
  }
}

/**
 * Find the path to loza-js.defaults.yaml.
 * Search order: LOZA_JS_DEFAULTS env, cwd, package root.
 */
function findDefaultsPath(): string | null {
  const envOverride = process.env.LOZA_JS_DEFAULTS?.trim();
  if (envOverride && existsSync(envOverride)) return envOverride;

  const cwdCandidate = join(process.cwd(), 'loza-js.defaults.yaml');
  if (existsSync(cwdCandidate)) return cwdCandidate;

  const pkgCandidate = join(getPackageRoot(), 'loza-js.defaults.yaml');
  if (existsSync(pkgCandidate)) return pkgCandidate;

  return null;
}

/**
 * Find the path to user config file.
 * Search order: LOZA_JS_CONFIG env, .loza-js.yaml in cwd, loza.yaml in cwd.
 */
function findUserConfigPath(): string | null {
  const envOverride = process.env.LOZA_JS_CONFIG?.trim();
  if (envOverride && existsSync(envOverride)) return envOverride;

  const dotCandidate = join(process.cwd(), '.loza-js.yaml');
  if (existsSync(dotCandidate)) return dotCandidate;

  const lozaCandidate = join(process.cwd(), 'loza.yaml');
  if (existsSync(lozaCandidate)) return lozaCandidate;

  return null;
}

// ── Overlay ─────────────────────────────────────────────────────────────────

/** Deep-merge override onto base (override wins for non-empty values). */
function overlayRaw(base: YamlRecord, override: YamlRecord): YamlRecord {
  const result: YamlRecord = { ...base };
  for (const [key, value] of Object.entries(override)) {
    if (value === undefined || value === null || value === '') continue;
    if (
      typeof value === 'object' &&
      !Array.isArray(value) &&
      typeof result[key] === 'object' &&
      result[key] !== null &&
      !Array.isArray(result[key])
    ) {
      result[key] = overlayRaw(
        result[key] as YamlRecord,
        value as YamlRecord,
      );
    } else {
      result[key] = value;
    }
  }
  return result;
}

// ── Public API ──────────────────────────────────────────────────────────────

/**
 * Load and overlay YAML config files (defaults + user override).
 * Returns the merged raw YAML dict, or empty object if no files are found.
 * Only works in Node.js environments (silently returns {} in browsers).
 */
export function loadFileConfig(): YamlRecord {
  if (typeof process === 'undefined') return {};

  let raw: YamlRecord = {};

  // Layer 1: SDK defaults
  const defaultsPath = findDefaultsPath();
  if (defaultsPath) {
    try {
      const content = readFileSync(defaultsPath, 'utf-8');
      raw = parseSimpleYAML(content);
    } catch {
      // Read failed, continue with empty
    }
  }

  // Layer 2: User overrides (overlay on top of defaults)
  const userPath = findUserConfigPath();
  if (userPath) {
    try {
      const content = readFileSync(userPath, 'utf-8');
      const userRaw = parseSimpleYAML(content);
      raw = overlayRaw(raw, userRaw);
    } catch {
      // Read failed, continue with defaults only
    }
  }

  return raw;
}

/**
 * Apply raw YAML config values to a Config object.
 * Only non-empty YAML values are applied (empty/missing fields are skipped).
 * If collector_url is a loza:// DSN, it is parsed and resolved to an HTTP URL.
 */
export function mergeFileConfig(base: Config, raw: YamlRecord): Config {
  const cfg = { ...base };

  // ── Top-level scalar fields ────────────────────────────────────────────────

  applyString(raw, cfg, 'collector_url', 'collectorUrl');
  applyString(raw, cfg, 'service_name', 'service');
  applyString(raw, cfg, 'service', 'service');
  applyString(raw, cfg, 'service_version', 'version');
  applyString(raw, cfg, 'version', 'version');
  applyString(raw, cfg, 'environment', 'environment');
  applyString(raw, cfg, 'release', 'release');
  applyString(raw, cfg, 'namespace', 'namespace');
  applyString(raw, cfg, 'api_key', 'apiKey');
  applyString(raw, cfg, 'level', 'level');
  applyString(raw, cfg, 'duplicate_policy', 'duplicatePolicy');

  applyNumber(raw, cfg, 'batch_size', 'batchSize');
  applyNumber(raw, cfg, 'flush_interval', 'flushIntervalMs');
  // max_buffer_size maps to async.queueSize (applied after nested section)
  applyNumber(raw, cfg, 'max_retries', 'maxRetries');
  applyNumber(raw, cfg, 'max_backoff', 'maxBackoffMs');
  applyNumber(raw, cfg, 'timeout', 'timeoutMs');

  applyBoolean(raw, cfg, 'strict', 'strict');
  applyBoolean(raw, cfg, 'enable_compression', 'enableCompression');
  applyBoolean(raw, cfg, 'include_host', 'includeHost');
  applyBoolean(raw, cfg, 'include_runtime', 'includeRuntime');

  // ── Parse collector_url as DSN if it uses loza:// scheme ───────────────────

  if (cfg.collectorUrl && cfg.collectorUrl.startsWith('loza://')) {
    try {
      const dsn = parseDSN(cfg.collectorUrl);
      cfg.collectorUrl = dsn.baseURL;
      cfg.collectorName = dsn.collectorName;
      if (dsn.username !== undefined) {
        cfg.username = dsn.username;
        cfg.password = dsn.password ?? '';
      }
      // Extract DSN-derived values (only override if not already set from YAML)
      if (dsn.env && dsn.env !== 'default' && cfg.environment === 'development') {
        cfg.environment = dsn.env;
      }
      if (dsn.service && !cfg.service) {
        cfg.service = dsn.service;
      }
    } catch {
      // Invalid DSN — leave as-is, env vars or code config may fix it
    }
  }

  // ── Nested: async ──────────────────────────────────────────────────────────

  const asyncRaw = raw.async ?? raw.async_config;
  if (asyncRaw && typeof asyncRaw === 'object' && !Array.isArray(asyncRaw)) {
    const a = asyncRaw as YamlRecord;
    const async = { ...cfg.async };
    if (a.enabled !== undefined) async.enabled = Boolean(a.enabled);
    if (a.queue_size !== undefined && typeof a.queue_size === 'number') async.queueSize = a.queue_size;
    if (a.queueSize !== undefined && typeof a.queueSize === 'number') async.queueSize = a.queueSize;
    if (a.workers !== undefined && typeof a.workers === 'number') async.workers = a.workers;
    if (a.max_batch_bytes !== undefined && typeof a.max_batch_bytes === 'number') async.maxBatchBytes = a.max_batch_bytes;
    if (a.maxBatchBytes !== undefined && typeof a.maxBatchBytes === 'number') async.maxBatchBytes = a.maxBatchBytes;
    if (a.flush_interval_ms !== undefined && typeof a.flush_interval_ms === 'number') async.flushIntervalMs = a.flush_interval_ms;
    if (a.flushIntervalMs !== undefined && typeof a.flushIntervalMs === 'number') async.flushIntervalMs = a.flushIntervalMs;
    cfg.async = async;
  }

  // ── Nested: security ──────────────────────────────────────────────────────

  const secRaw = raw.security;
  if (secRaw && typeof secRaw === 'object' && !Array.isArray(secRaw)) {
    const s = secRaw as YamlRecord;
    const security = { ...cfg.security };
    if (s.redact_by_default !== undefined) security.redactByDefault = Boolean(s.redact_by_default);
    if (s.redactByDefault !== undefined) security.redactByDefault = Boolean(s.redactByDefault);
    if (s.allow_pii !== undefined) security.allowPII = Boolean(s.allow_pii);
    if (s.allowPII !== undefined) security.allowPII = Boolean(s.allowPII);
    if (s.max_field_bytes !== undefined && typeof s.max_field_bytes === 'number') security.maxFieldBytes = s.max_field_bytes;
    if (s.maxFieldBytes !== undefined && typeof s.maxFieldBytes === 'number') security.maxFieldBytes = s.maxFieldBytes;
    if (s.max_event_bytes !== undefined && typeof s.max_event_bytes === 'number') security.maxEventBytes = s.max_event_bytes;
    if (s.maxEventBytes !== undefined && typeof s.maxEventBytes === 'number') security.maxEventBytes = s.maxEventBytes;
    if (s.max_attr_count !== undefined && typeof s.max_attr_count === 'number') security.maxAttrCount = s.max_attr_count;
    if (s.maxAttrCount !== undefined && typeof s.maxAttrCount === 'number') security.maxAttrCount = s.maxAttrCount;
    if (s.drop_oversized_events !== undefined) security.dropOversizedEvents = Boolean(s.drop_oversized_events);
    if (s.dropOversizedEvents !== undefined) security.dropOversizedEvents = Boolean(s.dropOversizedEvents);
    cfg.security = security;
  }

  // ── Top-level max_buffer_size -> async.queueSize ───────────────────────────
  // The loza-js.defaults.yaml uses max_buffer_size at the top level,
  // which maps to async.queueSize in the JS Config interface.

  const maxBuf = raw.max_buffer_size;
  if (typeof maxBuf === 'number' && !isNaN(maxBuf)) {
    cfg.async = { ...cfg.async, queueSize: maxBuf };
  }

  return cfg;
}

// ── Helpers ─────────────────────────────────────────────────────────────────

/** Apply a string YAML value to a Config field. */
function applyString(
  raw: YamlRecord,
  cfg: Config,
  yamlKey: string,
  cfgKey: keyof Config,
): void {
  const value = raw[yamlKey];
  if (typeof value === 'string' && value !== '') {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (cfg as any)[cfgKey] = value;
  }
}

/** Apply a numeric YAML value to a Config field. */
function applyNumber(
  raw: YamlRecord,
  cfg: Config,
  yamlKey: string,
  cfgKey: keyof Config,
): void {
  const value = raw[yamlKey];
  if (typeof value === 'number' && !isNaN(value)) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (cfg as any)[cfgKey] = value;
  }
}

/** Apply a boolean YAML value to a Config field. */
function applyBoolean(
  raw: YamlRecord,
  cfg: Config,
  yamlKey: string,
  cfgKey: keyof Config,
): void {
  const value = raw[yamlKey];
  if (typeof value === 'boolean') {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (cfg as any)[cfgKey] = value;
  }
}
